tools/scripts/docs_modm_io_generator.py

```diff
@@ -0,0 +1,403 @@
+			targetinput.style.borderColor = "#fff";
+		}, 5000);
+		return;
+	}
+	if(!releaseinput.value) {
+		releaseinput.style.transition = "border 5ms ease-out";
+		releaseinput.style.borderColor = "red";
+		setTimeout(function(){
+			releaseinput.style.transition = "border 5000ms ease-out";
+			releaseinput.style.borderColor = "#fff";
+		}, 5000);
+		return;
+	}
+	var url = "/" + releaseinput.value + "/api/" + targetinput.value + "/";
+	location.href = url;
+}
+targetinput.addEventListener("input", function(event) {
+	var a, b, i, val = this.value;
+	closeAllLists();
+	if (!val) { return false;}
+	currentFocus = -1;
+	a = document.createElement("DIV");
+	a.setAttribute("id", this.id + "autocomplete-list");
+	a.setAttribute("class", "autocomplete-items");
+	this.parentNode.appendChild(a);
+	for (i = 0; i < devices.length; i++) {
+		var regex = RegExp(val.toLowerCase());
+		if (devices[i].toLowerCase().match(regex) != null) {
+			b = document.createElement("DIV");
+			b.innerHTML = devices[i].replace(regex, function(str) { return "<strong>" + str + "</strong>" })
+			b.innerHTML += "<input type='hidden' value='" + devices[i] + "'>";
+			b.addEventListener("click", function(event) {
+				targetinput.value = this.getElementsByTagName("input")[0].value;
+				closeAllLists();
+			});
+			a.appendChild(b);
+		}
+	}
+});
+targetinput.addEventListener("keydown", function(event) {
+	var list = document.getElementById(this.id + "autocomplete-list");
+	if (list) {
+		list = list.getElementsByTagName("div");
+	}
+	if (event.keyCode == 40) { // down key
+		currentFocus++;
+		makeItemActive(list);
+	} else if (event.keyCode == 38) { // up key
+		currentFocus--;
+		makeItemActive(list);
+	} else if (event.keyCode == 13) { // enter key
+		event.preventDefault();
+		if (currentFocus > -1) {
+			/*and simulate a click on the "active" item:*/
+			if (list) {
+				list[currentFocus].click();
+			}
+		}
+	}
+});
+function removeActiveFromList(list) {
+	for (var i = 0; i < list.length; i++) {
+		list[i].classList.remove("autocomplete-active");
+	}
+}
+function makeItemActive(list) {
+	if (!list) return false;
+	removeActiveFromList(list);
+	if (currentFocus >= list.length) {
+		currentFocus = 0;
+	}
+	if (currentFocus < 0) {
+		currentFocus = (list.length - 1);
+	}
+	list[currentFocus].classList.add("autocomplete-active");
+}
+function closeAllLists(element) {
+	var items = document.getElementsByClassName("autocomplete-items");
+	for (var i = 0; i < items.length; i++) {
+		if (element != items[i] && element != targetinput) {
+			items[i].parentNode.removeChild(items[i]);
+		}
+	}
+}
+document.addEventListener("click", function (element) {
+    closeAllLists(element.target);
+});
+</script>
+<noscript>
+<h3>Select your target</h3>
+<input type="text" id="input" placeholder="Type e.g. 'F407'" />
+<ul>
+{% for d in devices %}
+ <li id="{{ d }}"><a href="/develop/api/{{ d }}/">{{ d }}</a></li>
+{% endfor %}
+</ul>
+</noscript>
+</main>
+</body>
+</html>
+"""
```

Comment:
Please put this into a separate .html.in file, this will likely grow even longer in future.

Reply:
I want to keep the script self-contained because it makes document generation on the server side more flexible.
The server can simply do the following:
```
wget -O - https://raw.githubusercontent.com/modm-io/modm/develop/tools/scripts/docs_modm_io_generator.py | python3
```

Reply:
But you need to clone the repository anyways, I don't see how this is better than `git clone modm && python3 modm/tools/scripts/foo.py`.

Reply:
I don't like the idea of automatically running a script downloaded from the Internet and would not do that on my server or on the roboterclubaachens server.

On an isolated VM (like CI) this would be a different matter, but it is problematic to save/transmit the output there and to execute the re-generation regularly.

Reply:
And yet you're executing the modm lbuild modules which are just plain Python scripts downloaded from the internet…

Reply:
Good argument.
I will change the script to use the repo instead of cloning it to temporary directory and put the template into a separate file.

I will add some proper sand-boxing (chroot, cgroups, ...) on the server and add a separate helper script.